### PR TITLE
PC-958: Init

### PR DESCRIPTION
**Pull request by ulfryk**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,63 @@
-# search-ql
-QL used for advanced searches in Samwise applications
+# SearchQL
+
+QL used for advanced searches in in Samwise applications. It's a slightly updated subset of [GitHub search syntax][1]. Implemented using [parsimmon][2] - a monadic parser combinator library for javascript.
```
Comment (tomecko) [inline README.md]: :eye:

File: README.md [hunk 2]
```diff
@@ -1,2 +1,63 @@
...
+
+## Syntax
+
+There are few special words and characters used in SearchQL (configurable):
+- logical operators `AND`, `OR` and `NOT` (* `NOT` is not implemented yet)
+- logical operators `AND`, `OR` and `NOT` (* `NOT` is not implemented yet)
```
Comment (tomecko) [inline README.md]: `NOT` has been implemented

File: README.md [hunk 3]
```diff
@@ -1,2 +1,63 @@
...
+### Query expressions
+
+#### Basic expression
+
+The simplest expression is just one `word` without any special SearchQL syntax elements. If query starts and ends with `"` and has no more `"` inside, text between `"` and `"` will be treated as search text even if it contains any special words or chars.
+
+- `ispum`
+- `"Lorem, ispum: dolor AND sit amet"`
+
+**SPECIAL CASE** - if no syntax words or signs appear, whole string will be used as output. This case will not work for sub queries described below.
```
Comment (tomecko) [inline README.md]: > will be used as output
this may not been comprehensible

File: README.md [hunk 4]
```diff
@@ -1,2 +1,63 @@
...
+- `("delivery" AND NOT date: (jun OR jul OR apri)) OR status: pend`
+- `desc: (similarities OR "Lorem ipsum dolor sit")`
+
+### Query expressions
+
+#### Basic expression
+
+The simplest expression is just one `word` without any special SearchQL syntax elements. If query starts and ends with `"` and has no more `"` inside, text between `"` and `"` will be treated as search text even if it contains any special words or chars.
+
+- `ispum`
```
Comment (tomecko) [inline README.md]: there are people who tend to notice even such small typos :D

File: package.json [hunk 5]
```diff
@@ -0,0 +1,70 @@
+{
+  "name": "@samwise-tech/search-ql",
+  "version": "0.0.2",
+  "description": "QL used for advanced searches in in Samwise applications",
```
Comment (tomecko) [inline package.json]: in in :eye:

File: package.json [hunk 6]
```diff
@@ -0,0 +1,70 @@
+{
+  "name": "@samwise-tech/search-ql",
+  "version": "0.0.2",
```
Comment (tomecko) [inline package.json]: I've been thinking about it...
semver.org says:
> The simplest thing to do is start your initial development release at 0.1.0 and then increment the minor version for each subsequent release.

;)

File: README.md (hunk 1 above)
Comment (ulfryk) [reply README.md]: ?

Comment (tomecko) [reply README.md]: double "in" word

**Comment by ulfryk** [comment]
https://github.com/samwise-tech/search-ql/blob/feature/PC-958-init/README.md

File: README.md [hunk 7]
```diff
@@ -1,2 +1,148 @@
...
+// list of values or for building queries that work with other systems (e.g. ElasticSearch)
+const validExpression = parse('desc: (ipsum OR NOT dolor) AND "john doe"');
+const failure = parse('des:: c: ipsum OR NOT dolor AND "john doe":"');
+
+/* 
+ * The returned expression is wrapped in a Either monad (coming from [monet][3]), it's 
+ * Right { Expression } if passed string was parsable or Left { ParseFailure } if parser
+ * failed to parse given query.
+
+validExpression.forEach(expression => console.log(expression)); =>
```
Comment (tomecko) [inline README.md]: :eye:

File: README.md [hunk 8]
```diff
@@ -1,2 +1,148 @@
...
+
+JoinedExpression {
+  type: AND,
+  value: Set {
+    LabelledExpression {
+      label: "desc"
```
Comment (tomecko) [inline README.md]: missing `,` or maybe I don't know the syntax of this language ;)

File: README.md [hunk 9]
```diff
@@ -1,2 +1,148 @@
...
+          BasicExpression { "ipsum" },
+          NotExpression { BasicExpression { "dolor" } }
+        }
+      }
+    },
+    BasicExpression { "john doe" }
+  }
+}
+
+failure.forEach(err => console.log(err)); =>
```
Comment (tomecko) [inline README.md]: :eye:

File: README.md [hunk 10]
```diff
@@ -1,2 +1,148 @@
...
+- `desc: (ipsum OR dolor)`
+- `desc: NOT dolor`
+- `description: ("Lorem, ipsum: amet" AND NOT barbecue)`
+
+### Additional rules (to consider)
+
+- **?** `AND` can be replaced with `&` or ` ` (whitespace)
+- **?** `OR` can be replaced with `,` or `|`
+- **?** `NOT` can be replaced with `!`
+- ~`:` preceded with whitespace makes query following it a `Bascic query`~
```
Comment (tomecko) [inline README.md]: Basic

File: README.md [hunk 11]
```diff
@@ -1,2 +1,148 @@
...
+
+## Usage
+
+```typescript
+import { Map } from 'immutable';
+import { parseSearchQL, ParserName } from './src';
+
+// Yeas, it's configurable, choose what you need!
```
Comment (tomecko) [inline README.md]: Yeah ;)

**Pull request merged**